src/shared/src/types.rs

```diff
@@ -100,3 +100,57 @@ pub mod custom_token {
         Icrc(LedgerId),
     }
 }
+
+/// Types specifics to the user profile.
+pub mod user_profile {
+    use candid::{CandidType, Deserialize, Principal};
+    use std::collections::BTreeMap;
+
+    pub type CredentialType = String;
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct UserCredential {
+        pub credential_type: CredentialType,
+        pub verified_date_timestamp: Option<u64>,
+        pub expire_date_timestamp: Option<u64>,
+    }
+
+    // Used in the endpoint
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct UserProfile {
+        pub credentials: Vec<UserCredential>,
+        pub created_timestamp: u64,
+        pub updated_timestamp: u64,
+    }
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct UserProfileStorage {
+        pub credentials: BTreeMap<CredentialType, UserCredential>,
+        pub created_timestamp: u64,
+        pub updated_timestamp: u64,
+    }
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct AddCredentialRequest {
+        pub credential_jwt: String,
+    }
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct GetUsersRequest {
+        pub updated_after_timestamp: Option<u64>,
+        pub limit_response: Option<u64>,
+    }
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct OisyUser {
+        pub principal: Principal,
+        pub pouh_verified: bool,
+    }
+
+    #[derive(CandidType, Deserialize, Clone, Eq, PartialEq, Debug)]
+    pub struct GetUsersResponse {
```

Comment:
Sharing in case interesint, that's the result interface of Juno's satellite:

```
#[derive(Default, CandidType, Deserialize, Clone)]
    pub struct ListResults<T> {
        pub items: Vec<(Key, T)>,
        pub items_length: usize,
        pub items_page: Option<usize>,
        pub matches_length: usize,
        pub matches_pages: Option<usize>,
    }
```

Docs:

```
{
  items: []; // The data - array of documents
  items_length: bigint; // The number of documents - basically items.length
  items_page?: bigint; // If the query is paginated, at what page (starting from 0) do the items find the place
  matches_length: bigint; // The total number of matching results
  matches_pages?: bigint; // If the query is paginated, the total number (starting from 0) of pages
}
```

If we aim for pagination thought it might be interesting to share.

Reply:
That's very interesting. We might end up adding those. This will be used by the Rewards Canister, so we'll try to keep it simple.

Reply:
Sure. Reading the struct again:

- Shouldn't `data` be renamed to `users`?
- What's `limit_response` ? Feels like a call to action to prepare the response but, here it's already the response

Reply:
It's to tell the client what's the maximum size it will bet. Most of the times, it will be the same as in the request, except if the backend limit is 1,000 and the client asks for 50,000. In that case, the `limit_response` will be different in the request and response. Request is `50,000` and response is `1,000`.

Maybe the name is confusing. Therefore, I renamed to `matches_max_length`.

Reply:
Thanks